backend/rust/models/src/whisper/utils.rs

```diff
@@ -51,3 +62,359 @@ pub fn attn_decoder_mask<B: Backend>(seq_length: usize) -> Tensor<B, 2> {
+        -Self::tensor_max(-x, -min)
+    }
+
+    pub fn tensor_log10<B: Backend, const D: usize>(x: Tensor<B, D>) -> Tensor<B, D> {
+        let ln10 = (10.0f64).ln();
+        x.log() / ln10
+    }
+
+    pub fn all_zeros<B: Backend, const D: usize>(x: Tensor<B, D>) -> bool {
+        x.abs().max().into_scalar().to_f64().unwrap() == 0.0
+    }
+
+    pub fn _10pow<B: Backend, const D: usize>(x: Tensor<B, D>) -> Tensor<B, D> {
+        let log10 = (10.0f64).ln();
+        (x * log10).exp()
+    }
+
+    pub fn reverse<B: Backend, const D: usize, K: TensorKind<B> + BasicOps<B> + Numeric<B>>(
+        x: Tensor<B, D, K>,
+        dim: usize,
+    ) -> Tensor<B, D, K>
+    where
+        <K as BasicOps<B>>::Elem: Element,
+    {
+        let len = x.dims()[dim];
+        let indices = -Tensor::arange_device(0..len, &x.device()) + (len - 1) as i64;
+        x.select(dim, indices)
+    }
+}
+
+pub struct Audio {}
+
+impl Audio {
+    /// Returns the maximum number of waveform samples that can be sumbitted to `prep_audio` without
+    /// receiving more than `n_frame_max` frames.
+    pub fn max_waveform_samples(n_frame_max: usize) -> usize {
+        // The number of waveform samples must be less than this
+        let n_samples_max = HOP_LENGTH * (n_frame_max + 1) + Self::is_odd(N_FFT) as usize;
+        n_samples_max - 1
+    }
+
+    fn is_odd(x: usize) -> bool {
+        if x % 2 == 0 {
+            false
+        } else {
+            true
+        }
+    }
+
+    pub fn fft_frequencies_device<B: Backend>(
+        sample_rate: f64,
+        n_fft: usize,
+        device: &B::Device,
+    ) -> Tensor<B, 1> {
+        Tensor::arange_device(0..(n_fft / 2 + 1), device)
+            .float()
+            .mul_scalar(sample_rate / n_fft as f64)
+    }
+    pub fn hz_to_mel(freq: f64, htk: bool) -> f64 {
+        if htk {
+            return 2595.0 * (1.0 + freq / 700.0).log10();
+        }
+
+        // Fill in the linear part
+        let f_min = 0.0;
+        let f_sp = 200.0 / 3.0;
+
+        // File in the log-scale part
+
+        let min_log_hz = 1000.0; // beginning of log region (Hz)
+        let min_log_mel = (min_log_hz - f_min) / f_sp;
+        let logstep = (6.4f64).ln() / 27.0; // step size for log region
+
+        let mel = if freq >= min_log_hz {
+            min_log_mel + (freq / min_log_hz).ln() / logstep
+        } else {
+            (freq - f_min) / f_sp
+        };
+
+        return mel;
+    }
+
+    pub fn mel_to_hz_tensor<B: Backend>(mel: Tensor<B, 1>, htk: bool) -> Tensor<B, 1> {
+        if htk {
+            return (Helper::_10pow(mel / 2595.0) - 1.0) * 700.0;
+        }
+
+        // Fill in the linear scale
+        let f_min = 0.0;
+        let f_sp = 200.0 / 3.0;
+
+        // And now the nonlinear scale
+        let min_log_hz = 1000.0; // beginning of log region (Hz)
+        let min_log_mel = (min_log_hz - f_min) / f_sp;
+        let logstep = (6.4f64).ln() / 27.0; // step size for log region
```

Comment:
those constants are repeated, being always f64 you can just keep them as `const`s

Reply:
thank you, will do.